src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -111,6 +111,15 @@ public List<TSource> ToList()
                 return list;
             }
 
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                List<TSource> list = _source.ToList();
+
+                list.Reverse();
+
+                return new HashSet<TSource>(list, comparer);
```

Comment:
Yes, this should definitely skip the reversing. (This and the ordering are case where there are a lot of potential optimisations we could do by just skipping the reversing or ordering operation, particularly with all the aggregate operations, but the cost of testing for that condition isn't justified when the caller should just do that skipping anyway. When we're going through IListProvider to optimise other cases we've already paid that cost, as with `Count()`).